File: client/src/app/verify/framing.ts
/**
 * Unambiguous hashing of several fields — the TypeScript half of `server/src/framing.rs`.
 *
 * Plain concatenation is ambiguous whenever a field's length can vary: ("ab","c") and ("a","bc")
 * hash identically. Every field is length-prefixed instead, with the same rule the pool manifest
 * hash uses, so there is one thing to reason about rather than a case analysis.
 */

/** 64-bit little-endian, matching Rust's `u64::to_le_bytes`. */
export function le64(n: bigint): Uint8Array {
  const out = new Uint8Array(8);
  let v = n;
  for (let i = 0; i < 8; i++) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

export function utf8(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

export function toHex(b: Uint8Array): string {
  return Array.from(b)
    .map((x) => x.toString(16).padStart(2, '0'))
    .join('');
}

export function fromHex(s: string): Uint8Array {
  if (s.length % 2 !== 0) throw new Error(`hex string has odd length: ${s}`);
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(s.substr(i * 2, 2), 16);
  return out;
}

/**
 * `LE64(|p0|) ‖ p0 ‖ LE64(|p1|) ‖ p1 ‖ …` — the bytes that actually go into SHA-256.
 *
 * Split out from `framed` so the interface can show a reader the exact preimage rather than
 * asking them to take the digest on faith. There is still one definition of the layout.
 */
export function frame(parts: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const p of parts) total += 8 + p.length;
  const buf = new Uint8Array(total);
  let o = 0;
  for (const p of parts) {
    buf.set(le64(BigInt(p.length)), o);
    o += 8;
    buf.set(p, o);
    o += p.length;
  }
  return buf;
}

/** `SHA-256( LE64(|p0|) ‖ p0 ‖ LE64(|p1|) ‖ p1 ‖ … )` */
export async function framed(parts: Uint8Array[]): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.digest('SHA-256', frame(parts)));
}

/** The `sha256:`-prefixed form used in the API and the public log. */
export async function framedHex(parts: Uint8Array[]): Promise<string> {
  return 'sha256:' + toHex(await framed(parts));
}

/**
 * The per-trial commitment (D3). The coordinate is inside the hash, which is what makes the
 * reveal prove *this* coordinate pointed at *this* image.
 */
export async function commitment(
  sServer: Uint8Array,
  nonce: Uint8Array,
  coordinate: string,
): Promise<string> {
  return framedHex([sServer, nonce, utf8(coordinate)]);
}

File: server/src/framing.rs
//! Unambiguous hashing of several fields.
//!
//! Plain concatenation is ambiguous whenever a field's length can vary: `("ab","c")` and
//! `("a","bc")` hash identically. The commitment originally used plain concatenation and was
//! safe only because seeds are 32 bytes and the coordinate is `NNNN-NNNN` — an argument that
//! silently expires the day either format changes. Every field is length-prefixed instead.
//!
//! The same scheme is used for the pool manifest hash, so there is one rule to reason about.

use sha2::{Digest, Sha256};

/// `SHA-256( LE64(len₀) ‖ part₀ ‖ LE64(len₁) ‖ part₁ ‖ … )`
pub fn framed(parts: &[&[u8]]) -> [u8; 32] {
    let mut h = Sha256::new();
    for p in parts {
        h.update((p.len() as u64).to_le_bytes());
        h.update(p);
    }
    h.finalize().into()
}

/// The same digest as a `sha256:`-prefixed lowercase hex string, which is how hashes appear in
/// the public log and in the API.
pub fn framed_hex(parts: &[&[u8]]) -> String {
    let d = framed(parts);
    let mut s = String::with_capacity(7 + 64);
    s.push_str("sha256:");
    for b in d {
        s.push_str(&format!("{b:02x}"));
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The defect this module exists to remove.
    #[test]
    fn field_boundaries_cannot_be_shifted() {
        assert_ne!(
            framed(&[b"ab", b"c", b"d"]),
            framed(&[b"a", b"bc", b"d"]),
            "length prefixes must make the split unambiguous"
        );
    }

    #[test]
    fn empty_fields_are_distinguishable() {
        assert_ne!(framed(&[b"", b"x"]), framed(&[b"x", b""]));
        assert_ne!(framed(&[b"x"]), framed(&[b"x", b""]));
    }

    #[test]
    fn hex_form_matches_the_digest() {
        let h = framed_hex(&[b"a"]);
        assert!(h.starts_with("sha256:"));
        assert_eq!(h.len(), 7 + 64);
    }
}
